[Scheduler] add `RecurringMessage::getId()` and prevent duplicates

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | n/a
| License       | MIT
| Doc PR        | todo

This could allow running specific scheduled tasks manually (via CLI or GUI).

kbond committed:
[Scheduler] add `RecurringMessage::getId()` and prevent duplicates

kbond commented:
This change breaks the current test suite for scheduler. I'd like to get some feedback or alternative ideas before going further.

Some context on my requirement for this: In my app, I have an admin GUI that lists all available scheduled tasks. User's can drill into each task (hence the need for a unique ID) and see a run history which includes memory usage and duration. The schedule name and ID on the `ScheduledStamp` would allow connecting specific runs with a recurring message.

There's likely other methods I can use to achieve this same system with symfony/scheduler so I'm not dead set on this being added.

kbond commented:
> There's likely other methods I can use to achieve this same system with symfony/scheduler so I'm not dead set on this being added.

I've played a little with this. I could have my own custom logic to generate an identifier for a `RecurringMessage` and I can infer the scheduler name from `RecievedStamp::$transport` (`schedule_default` or `schedule_foo`).

But I think I'd need the `RecurringMessage` added to the `ScheduledStamp` or the Trigger object as a bare minimum in order to connect runs with a specific `RecurringMessage`.

Basically, I think we need more context added to the `ScheduledStamp` but it's not possible with the way it's currently coded.

fabpot commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -51,6 +53,21 @@ public static function trigger(TriggerInterface $trigger, object $message): self
+
+        return $this->id = sha1(implode('', [
+            $this->message instanceof \Stringable ? (string) $this->message : $this->message::class,
+            $this->trigger instanceof \Stringable ? (string) $this->trigger : $this->trigger::class,
```
I would always use the class name as the string representation can be the same for 2 triggers or messages.

kbond replied:
I think I changed to what you've asked.

fabpot commented on src/Symfony/Component/Scheduler/Schedule.php:
```diff
@@ -29,8 +30,13 @@ final class Schedule implements ScheduleProviderInterface
-        $this->messages = array_merge($this->messages, $messages);
+        foreach ([$message, ...$messages] as $m) {
+            if (isset($this->messages[$m->getId()])) {
+                throw new LogicException('Duplicated schedule message.');
```
If using class names, I don't think we can have duplicates.

kbond replied:
I'm not sure what you mean here:
```php
$schedule = new Schedule();
$schedule->add(RecurringMessage::cron('* * * * *', new \stdClass()));
$schedule->add(RecurringMessage::cron('* * * * *', new \stdClass())); // is duplicate
```

fabpot reviewed (changes requested):
Having more information in the stamp does not look good as the component should work without using Messenger.

kbond commented:
> Having more information in the stamp does not look good as the component should work without using Messenger.

Can save for a followup PR but I was thinking to change `MessageGeneratorInterface::getMessages(): iterable<object>` to: `MessageGeneratorInterface::getMessages(): iterable<RecurringMessage>`.

In the stand-alone `Scheduler`, we'd unwrap the message from `RecurringMessage`. In the transport, we could add the `RecurringMessage` to the stamp.

kbond commented on src/Symfony/Component/Scheduler/RecurringMessage.php:
```diff
@@ -51,6 +53,23 @@ public static function trigger(TriggerInterface $trigger, object $message): self
+            return $this->id;
+        }
+
+        return $this->id = hash('crc32c', implode('', [
```
I used this hash alg because it's nice and short (for possibly displaying in `debug:schedule` and using in a future `schedule:run` command). Not sure if we need to be concerned about collisions.

stof replied:
if you need a unique identifier for usage in `schedule:run`, you need to be concerned about collisions

kbond replied:
> Not sure if we need to be concerned about collisions.

I meant in a mathematical sense using `crc32c` - how probable is a collision?

vtsykun commented:
Correctly I understand that now I can not use the same messages with the different arguments?

```php
            ->add(RM::cron('*/6 * * * *', new Class1('aaa')))
            ->add(RM::cron('*/7 * * * *', new Class1('bbb')))
            ->add(RM::cron('*/7 * * * *',  new Class1('ccc')));
```
In this case I got an error `Duplicated schedule message`. - will I need to create `toString` to avoid duplication restriction?

kbond commented:
> Correctly I understand that now I can not use the same messages with the different arguments?

`CronTrigger` is now `\Stringable` so those would be unique. I'd like to make `TriggerInterface extend \Stringable`.

vtsykun commented:
> CronExpressionTrigger is now \Stringable so those would be unique

Not sure that will be good. it means that we will not be able to use two tasks if their cron trigger is same.

Maybe more simple use iterable id if need to debug or execute the specific tasks manually, like this

```
php bin/console debug:scheduler --execute-one=7
```
![image1](https://user-images.githubusercontent.com/21358010/228657483-aa1c6350-6990-42db-8b27-4a411024b92c.png)

kbond commented:
> it means that we will not be able to use two tasks if their cron trigger is same.

In that case, yes, you'd need to make your message stringable.

> Maybe more simple use iterable id if need to debug or execute the specific tasks manually, like this

Indeed, that could be an option. I'd like to track runs in the db so those id's changing whenever a schedule is updated wouldn't be desired. However, if the `ScheduledStamp` had more context, I could overcome this problem.

kbond commented:
Rebased and refactored to and added the ID to `MessageContext` and therefore the `ScheduledStamp`.

To address [this concern](https://github.com/symfony/symfony/pull/49838#issuecomment-1489218761), I now cast the message to a string if stringable, if not, serialize it, if it can't be serialized, ignore. That example now works without needing `Class1` to be stringable.

fabpot commented:
@kbond Can you rebase and have a look at the failing tests?

kbond commented:
> Can you rebase and have a look at the failing tests?

Done.

fabpot commented:
Thank you @kbond.
